src/foam/nanos/auth/ServiceProviderAwareDAO.js

```diff
@@ -49,13 +66,48 @@ foam.CLASS({
         }
       }
     } else if ( ! sp.getSpid().equals(oldSp.getSpid()) &&
-                ! ((AuthService) x.get("auth")).check(x, "spid.update.*") ) {
+                ! (((AuthService) x.get("auth")).check(x, "spid.update."+oldSp.getSpid()) &&
+                   ((AuthService) x.get("auth")).check(x, "spid.update."+sp.getSpid())) ) {
       throw new AuthorizationException("You do not have permission to update ServiceProvider (spid) property.");
     }
 
     return super.put_(x, obj);
       `
+    },
+    {
+      name: 'find_',
+      javaCode: `
+    FObject result = getDelegate().find_(x, id);
+    if ( result == null ||
+         ((AuthService) x.get("auth")).check(x, "*") ) {
+      return result;
+    }
+
+    if ( new ServiceProviderAwareSupport().match(x, getPropertyInfos(), result) ) {
+      return result;
     }
-  ]
 
+    return null;
+      `
+    },
+    {
+      name: 'select_',
+      javaCode: `
+    if ( ((AuthService) x.get("auth")).check(x, "*") ) {
+      return super.select_(x, sink, skip, limit, order, predicate);
+    }
+
+    ProxySink proxy = (ProxySink) getDelegate().select_(
+      x,
+      new ServiceProviderAwareSink(x, sink, getPropertyInfos()),
```

Comment:
Hmm, I don't think this plays nicely with the other sink decorators. When `SkipSink`, `LimitSink`, `PredicatedSink`, and `OrderedSink` are added, they'll be _outside_ of this sink if I'm not mistaken. This is an issue because if we apply skip and limit before this sink gets hit, then we'll only be filtering out the limited number. What we really want is for the SPID filtering to occur before or as part of the predicate stage of the sink decorator chain, which happens before skip, limit, and order.